docs/architecture/adr-047-handling-evidence-from-light-client.md

```diff
@@ -0,0 +1,177 @@
+
+If the light client is under attack (either directly - lunatic / phantom
+validators or indirectly - fork on the main chain), it's supposed to halt and
+send evidence of misbehavior to the correct full node. Upon receiving an
+evidence, full node should punish malicious validators (if possible).
+
+## Decision
+
+When a light client sees two conflicting headers (`H1.Hash() != H2.Hash()`,
+`H1.Height == H2.Height`), both having 1/3+ of voting power of the currently
+trusted validator set, it will submit a `ConflictingHeadersEvidence` to all
+full nodes it's connected to.
+
+```go
+type ConflictingHeadersEvidence struct {
+  H1 types.SignedHeader
+  H2 types.SignedHeader
+}
+```
+
+When a full node receives the `ConflictingHeadersEvidence` evidence, it should
+a) validate it b) figure out if malicious behaviour is obvious (immediately
+slashable) or the fork accountability protocol needs to be started.
+
+### Validating headers
+
+Check both headers are valid (`ValidateBasic`), have
+the same height and signed by 1/3+ of known validator set.
+
+- Q: What if light client validator set is not equal to full node's validator
+  set (i.e. from full node's point of view both headers are not properly signed;
+  this includes the case where none of the two headers were committed on the
+  main chain)
+
+  Reject the evidence. It means light client is following a fork, but, hey, at
+  least it will halt.
+
+- Q: Don't we want to punish validators who signed something else even if they
+  have less or equal than 1/3?
+
+  No consensus so far. Ethan said no, Zarko said yes.
+  https://github.com/tendermint/spec/pull/71#discussion_r374210533
+
+### Figuring out if malicious behaviour is immediately slashable
+
+Let's say H1 was committed on the main chain. Intersect validator sets of H1
+and H2.
+
+* if there are signers(H2) that are not part of validators(H1), they misbehaved as
+they are signing protocol messages in heights they are not validators =>
+immediately slashable (#F4).
+
+* if `H1.Round == H2.Round`, and some signers signed different precommit
+messages in both commits, then it is an equivocation misbehavior => immediately
+slashable (#F1).
+
+* if `H1.Round != H2.Round` we need to run full detection procedure => not
+immediately slashable.
+
+* if `ValidatorsHash`, `NextValidatorsHash`, `ConsensusHash`,
+`AppHash`, and `LastResultsHash` in H2 are different (incorrect application
+state transition), then it is a lunatic misbehavior => immediately slashable (#F5).
+
+If evidence is not immediately slashable, fork accountability needs to invoked
+(ADR does not yet exist).
+
+It's unclear if we should further break up `ConflictingHeadersEvidence` or
+gossip and commit it directly. See
+https://github.com/tendermint/tendermint/issues/4182#issuecomment-590339233
+
+If we'd go without breaking evidence, all we'll need to do is to strip the
+committed header from `ConflictingHeadersEvidence` (H1) and leave only the
+uncommitted header (H2):
```

Comment:
Does the light client "remember" where it got the headers from? If H1 came from FN1 and H2 from FN2 the client submits evidence to both?

Reply:
to both

Reply:
here we're talking about a full node, not the light client! a full node needs to strip committed header (from it's perspective) and send only "malicious" header to other nodes.